python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+        return hostname, login, self._get_password()
+
+
+class ConsoleLoginHandler(ConsoleAuthenticationHandlerBase):
+    """
+    Handles username/password authentication.
+    """
+    def __init__(self, fixed_host):
+        super(ConsoleLoginHandler, self).__init__()
+        self._fixed_host = fixed_host
+
+    def _get_user_credentials(self, hostname, login):
+        """
+        Reads the user credentials from the keyboard.
+        :param hostname: Name of the host we will be logging on.
+        :param login: Default value for the login.
+        :returns: A tuple of (login, password) strings.
+        """
+        if self._fixed_host:
+            print "Please enter your login credentials for %s" % hostname
+        else:
+            print "Please enter your login credentials."
+            hostname = self._get_keyboard_input("Host", hostname)
+        login = self._get_keyboard_input("Login", login)
+        password = self._get_password()
+        return hostname, login, password
+
+
+class UiAuthenticationHandler(AuthenticationHandlerBase):
+    """
+    Handles ui based authentication.
+    """
+
+    def __init__(self, is_session_renewal, fixed_host=False):
+        """
+        Creates the UiAuthenticationHandler object.
+        :param is_session_renewal: Boolean indicating if we are renewing a session. True if we are, False otherwise.
+        """
+        self._is_session_renewal = is_session_renewal
+        self._gui_launcher = _create_invoker()
+        self._fixed_host = fixed_host
+
+    def authenticate(self, hostname, login, http_proxy):
+        """
+        Pops a dialog that asks for the hostname, login and password of the user. If there is a current
+        engine, it will run the code in the main thread.
+        :param hostname: Host to display in the dialog.
+        :param login: login to display in the dialog.
+        :param http_proxy: Proxy server to use when validating credentials. Can be None.
+        :returns: A tuple of (hostname, login, session_token)
+        """
+        from .ui import login_dialog
+
+        if self._is_session_renewal:
+            logger.debug("Requesting password in a dialog.")
+        else:
+            logger.debug("Requesting username and password in a dialog.")
+
+        def _process_ui():
+            dlg = login_dialog.LoginDialog(
+                "Shotgun Login",
+                is_session_renewal=self._is_session_renewal,
+                hostname=hostname,
+                login=login,
+                http_proxy=http_proxy,
+                fixed_host=self._fixed_host
+            )
+            return dlg.result()
+
+        result = self._gui_launcher(_process_ui)
+
+        if not result:
+            self.raise_authentication_cancelled_error()
+        return result
+
+
+_renew_session_lock = threading.Lock()
+"""
+Lock the assures only one thread at a time can execute the authentication logic.
```

Comment:
In other parts of the code, we have typically used comments for this sort of thing rather than docstrings.